### DISCUSSION TITLE
How to save and use the Client in AppState?

### DISCUSSION THREAD

**Discussion by InShade** [tokio-rs/axum#3092] [Q&A] [answered]
How can I reuse the client, i.e. reuse the underlying connection
in function `create_user`?

Trying to save the Client in AppState and use it so I don't have to create it again each time.

It doesn't work in my function. Want without sqlx.

How to do it correctly?
Please, fix my code. 
`

    #[tokio::main]
    async fn main() -> Result<(), Box<dyn std::error::Error>> {

       // POSTGRESQL - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       let (client, connection) = tokio_postgres::connect("host=localhost ***", NoTls).await?;
       tokio::spawn(async move {
        if let Err(e) = connection.await {
           eprintln!("connection error: {}", e);
        }
       });
       //
       let state = AppState {
          d: Arc::new(Mutex::new(vec![])),
          client: client
       };
       //
       // works without sqlx:
       let row = &client.query("insert into test (inf, res) values ($1,$2)", &[&"txt", &100]).await?;
       // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
   
       // AXUM  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
       // ...
       let app = Router::new()
        .route("/info", post(create_user))
        .with_state(state);

       let listener = tokio::net::TcpListener::bind("127.0.0.1:3000").await.unwrap();
       axum::serve(listener, app).await.unwrap();
       // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       // ...   
    Ok(())
    }

    async fn create_user(

        State(state): State<AppState>,
        Json(json): Json<Value>,
     ) -> Json<Value> {

        let mut s: String = "".to_string();
        // ...
    
        // Trying to save the Client in AppState and use it so I don't have to create it again each time.
        // doesn't work. want without sqlx :

        let row = &client.query("insert into test (inf, res) values ($1,$2)", &[&"abc", &10]).await.unwrap();

        // ...
        // println!("s = {:?}", &s);
        Json(serde_json::from_str::<serde_json::Value>(&s).expect("REASON"))
    }

    #[derive(Debug, Clone)]
    struct AppState {
        d: Arc<Mutex<Vec<String>>>,  // вектор строк 
       // client: reqwest::Client,
       client: tokio_postgres::Client,
    }
`

**Comment by mladedav** [comment]
You didn't provide any error messages or what specifically you have issues with so there isn't much help we can provide.

**Comment by yanns** [comment]
What is the issue if you save your client in `AppState`?
```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {

   // POSTGRESQL - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   let (client, connection) = tokio_postgres::connect("host=localhost ***", NoTls).await?;
   tokio::spawn(async move {
    if let Err(e) = connection.await {
       eprintln!("connection error: {}", e);
    }
   });

   // creates the client
   let created_client = &client.query("insert into test (inf, res) values ($1,$2)", &[&"txt", &100]).await?;

   //
   let state = AppState {
      d: Arc::new(Mutex::new(vec![])),
      client: client,
      created_client: created_client, // or Arc::new(created_client)
   };
   //
   // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

   // AXUM  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
   // ...
   let app = Router::new()
    .route("/info", post(create_user))
    .with_state(state);

   let listener = tokio::net::TcpListener::bind("127.0.0.1:3000").await.unwrap();
   axum::serve(listener, app).await.unwrap();
   // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   // ...   
Ok(())
}


async fn create_user(

    State(state): State<AppState>,
    Json(json): Json<Value>,
 ) -> Json<Value> {

    let mut s: String = "".to_string();
    // ...


    let row = state.created_client;

    // ...
    // println!("s = {:?}", &s);
    Json(serde_json::from_str::<serde_json::Value>(&s).expect("REASON"))
}


#[derive(Debug, Clone)]
struct AppState {
    d: Arc<Mutex<Vec<String>>>,  // вектор строк 
   // client: reqwest::Client,
   client: tokio_postgres::Client,
   created_client: <type of your created client>
}
```

**Comment by yanns** [comment]
What is the issue if you save your client in `AppState`?
```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {

   // POSTGRESQL - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   let (client, connection) = tokio_postgres::connect("host=localhost ***", NoTls).await?;
   tokio::spawn(async move {
    if let Err(e) = connection.await {
       eprintln!("connection error: {}", e);
    }
   });

   // creates the client
   let created_client = &client.query("insert into test (inf, res) values ($1,$2)", &[&"txt", &100]).await?;

   //
   let state = AppState {
      d: Arc::new(Mutex::new(vec![])),
      client: client,
      created_client: created_client, // or Arc::new(created_client)
   };
   //
   // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

   // AXUM  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
   // ...
   let app = Router::new()
    .route("/info", post(create_user))
    .with_state(state);

   let listener = tokio::net::TcpListener::bind("127.0.0.1:3000").await.unwrap();
   axum::serve(listener, app).await.unwrap();
   // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   // ...   
Ok(())
}


async fn create_user(

    State(state): State<AppState>,
    Json(json): Json<Value>,
 ) -> Json<Value> {

    let mut s: String = "".to_string();
    // ...


    let row = state.created_client;

    // ...
    // println!("s = {:?}", &s);
    Json(serde_json::from_str::<serde_json::Value>(&s).expect("REASON"))
}


#[derive(Debug, Clone)]
struct AppState {
    d: Arc<Mutex<Vec<String>>>,  // вектор строк 
   // client: reqwest::Client,
   client: tokio_postgres::Client,
   created_client: <type of your created client>
}
```

**Reply by InShade** [reply]
@yanns, thanks
if I do this, get error:

`

    error[E0277]: the trait bound `fn(axum::extract::State<AppState<'_>>, axum::Json<JsonValue>) -> impl Future<Output = 
    axum::Json<
    JsonValue>> {create_user}: Handler<_, _>` is not satisfied    
       --> src/main.rs:121:29
        |
    121 |       .route("/info", post(create_user))
        |                        ---- ^^^^^^^^^^^ the trait `Handler<_, _>` is not implemented for fn item `fn(axum::extract::State<
    AppState<'_>>, axum::Json<JsonValue>) -> impl Future<Output = axum::Json<JsonValue>> {create_user}`
        |                        |
        |                        required by a bound introduced by this call
        |
        = note: Consider using `#[axum::debug_handler]` to improve the error message
        = help: the following other types implement trait `Handler<T, S>`:
                  `MethodRouter<S>` implements `Handler<(), S>`
                  `axum::handler::Layered<L, H, T, S>` implements `Handler<T, S>`
    note: required by a bound in `post`
       --> ...\.cargo\registry\src\index.crates.io-6f17d22bba15001f\axum-0.7.9\src\routing\method_routing.rs:443:1
        |
    443 | top_level_handler_fn!(post, POST);
        | ^^^^^^^^^^^^^^^^^^^^^^----^^^^^^^
        | |                     |
        | |                     required by a bound in this function
        | required by this bound in `post`
        = note: this error originates in the macro `top_level_handler_fn` (in Nightly builds, run with -Z macro-backtrace for more i
    nfo)

`

![err_E0277_](https://github.com/user-attachments/assets/a815f869-c1a3-4e09-a31e-6030b70332b5)

~ code

`

        #[tokio::main]
        async fn main() -> Result<(), Box<dyn std::error::Error>> {

           // postgres 
           let (client, connection) = tokio_postgres::connect("host=localhost ***", NoTls).await?;
           tokio::spawn(async move {
              if let Err(e) = connection.await {
                 eprintln!("connection error: {}", e);
              }
           });
           let created_client = &client.query("insert into test (inf, res) values ($1,$2)", &[&"txt", &100]).await?;

           let state = AppState {
              // pool: pool,
              created_client: created_client, // or Arc::new(created_client)
           };
   
           // axum   
           let app = Router::new()
              // ... 
              .route("/info", post(create_user))
              .with_state(state);

           // hyper
           let listener = tokio::net::TcpListener::bind("127.0.0.1:3000").await.unwrap();
           axum::serve(listener, app).await.unwrap();

        Ok(())
        }


        async fn create_user(

           State(state): State<AppState>,
           Json(json): Json<Value>,
        ) -> Json<Value> {

            // let row = &client.query("insert into test (inf, res) values ($1,$2)", &[&"abc", &10]).await.unwrap();
            let row = state.created_client;

            // ...
        }


        #[derive(Debug, Clone)]
        struct AppState<'a> {
           //pool: PgPool,
           created_client: &'a Vec<Row>,  // &Vec<tokio_postgres::Row>
        }

`

#[debug_handler] ( is deprecated) 
vs 
axum-macros

Can it help in my case?

**Reply by yanns** [reply]
You can try the tip given by the compiler:
> note: Consider using `#[axum::debug_handler]` to improve the error message

You can find more info on https://docs.rs/axum/latest/axum/handler/index.html#debugging-handler-type-errors

**Reply by InShade** [reply]
@yanns

did not give any result,
it was suggested to do it like this _(error[E0106]: missing lifetime specifier):_
![err_E0106_](https://github.com/user-attachments/assets/db0c12a0-a4f2-41a4-9a99-8d7d4bfd09fd)

then like this _(error[E0782]: trait objects must include the `dyn` keyword)_:
` State(state): dyn for<'a> State<AppState<'a>>,`

now the only mistake _(error[E0404]: expected trait, found struct `State`)_:
![err_E0404_](https://github.com/user-attachments/assets/0869aa7d-a673-4be0-8f7b-0eb39b18e3ee)

**Reply by yanns** [reply]
Your `AppState` should have any lifetime.
Try this:
```rust
    #[derive(Debug, Clone)]
    struct AppState {
       //pool: PgPool,
       created_client: Vec<Row>,  // or Arc<Vec<Row>> 
    }
```

**Reply by InShade** [reply]
@yanns

        let state = AppState {
            // ...
            created_client: created_client.to_vec(),        //      or Arc::new(created_client)
                                           ^^^^^^^^_  without  `.to_vec()`  was forced to install with <'a>
        };

Thanks, the code variant

    async fn create_user(

       State(state): State<AppState>,
       Json(json): Json<Value>,
    ) -> Json<Value> {

        // let row = &client.query("insert into test (inf, res) values ($1,$2)", &[&"abc", &10]).await.unwrap();
        // let row = state.created_client;
        state.created_client = &client.query("insert into ... ").await.unwrap();
                                ^^^^^^___  error: not found in this scope    
        // ...
    }

When 

    #[derive(Debug, Clone)]
    struct AppState {
       // client: Client,
       // created_client: Vec<Row>,         // or Arc<Vec<Row>>
        client: tokio_postgres::Client,     // error[E0277]: the trait `Clone` is not implemented for `tokio_postgres::Client`
        created_client: Vec<tokio_postgres::row::Row>,
    }

    let state = AppState {
       // ...
       client: client,
       created_client: created_client.to_vec()
    };

gives an error:

error[E0277]: the trait `Clone` is not implemented for `tokio_postgres::Client`

**Reply by yanns** [reply, marked as answer]
yes, you can try to put it into an `Arc`.

**Reply by InShade** [reply]
Thank you!
